File: veriblock-core/src/main/java/org/veriblock/sdk/models/BitcoinBlock.java
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2019 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package org.veriblock.sdk.models;

import org.veriblock.core.crypto.Sha256Hash;
import org.veriblock.sdk.services.SerializeDeserializeService;

import java.util.Arrays;

public class BitcoinBlock {
    private final byte[] raw;
    private final int version;
    private final Sha256Hash previousBlock;
    private final Sha256Hash merkleRoot;
    private final int timestamp;
    private final int bits;
    private final int nonce;
    private final Sha256Hash hash;

    public BitcoinBlock(int version, Sha256Hash previousBlock, Sha256Hash merkleRoot, int timestamp, int bits, int nonce) {
        this.version = version;
        this.previousBlock = previousBlock;
        this.merkleRoot = merkleRoot;
        this.timestamp = timestamp;
        this.bits = bits;
        this.nonce = nonce;
        //TODO move it out of here.
        this.raw = SerializeDeserializeService.getHeaderBytesBitcoinBlock(this);
        this.hash = Sha256Hash.wrapReversed(Sha256Hash.hashTwice(this.raw));
    }

    @Override
    public boolean equals(Object o) {
        return this == o || o != null && getClass() == o.getClass() && Arrays.equals(SerializeDeserializeService.getHeaderBytesBitcoinBlock(this), SerializeDeserializeService
            .getHeaderBytesBitcoinBlock((BitcoinBlock) o));
    }

    public byte[] getRaw() {
        return raw;
    }

    public int getVersion() {
        return version;
    }

    public Sha256Hash getPreviousBlock() {
        return previousBlock;
    }

    public Sha256Hash getMerkleRoot() {
        return merkleRoot;
    }

    public Sha256Hash getMerkleRootReversed() {
        return Sha256Hash.wrap(merkleRoot.getReversedBytes());
    }

    public int getTimestamp() {
        return timestamp;
    }

    public int getBits() {
        return bits;
    }

    public int getNonce() {
        return nonce;
    }

    public Sha256Hash getHash() {
        return this.hash;
    }
}

File: veriblock-core/src/main/java/org/veriblock/sdk/models/BitcoinBlock.kt
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package org.veriblock.sdk.models

import org.veriblock.core.crypto.BtcHash
import org.veriblock.core.crypto.MerkleRoot
import org.veriblock.core.crypto.asMerkleRoot
import org.veriblock.core.crypto.btcHashOf
import org.veriblock.sdk.util.putLEBytes
import org.veriblock.sdk.util.putLEInt32
import java.nio.ByteBuffer
import java.util.Arrays

class BitcoinBlock(
    val version: Int,
    val previousBlock: BtcHash,
    val merkleRoot: MerkleRoot,
    val timestamp: Int,
    val difficulty: Int,
    val nonce: Int
) {
    val raw: ByteArray = run {
        val buffer = ByteBuffer.allocateDirect(Constants.HEADER_SIZE_BitcoinBlock)
        buffer.putLEInt32(version)
        buffer.putLEBytes(previousBlock.bytes)
        buffer.putLEBytes(merkleRoot.bytes)
        buffer.putLEInt32(timestamp)
        buffer.putLEInt32(difficulty)
        buffer.putLEInt32(nonce)
        buffer.flip()
        val bytes = ByteArray(Constants.HEADER_SIZE_BitcoinBlock)
        buffer[bytes]
        bytes
    }

    val hash: BtcHash = btcHashOf(raw)

    fun getMerkleRootReversed(): MerkleRoot =
        merkleRoot.reversedBytes().asMerkleRoot()

    override fun equals(other: Any?): Boolean {
        return this === other || other != null && javaClass == other.javaClass && Arrays.equals(
            raw, (other as BitcoinBlock).raw
        )
    }

    override fun hashCode(): Int {
        return raw.contentHashCode()
    }

    override fun toString(): String {
        return "BtcBlock(hash=${hash} prev=${previousBlock} time=${timestamp})"
    }
}
